### ISSUE TITLE
Enhancing the Flexibility of Linear Models in Leaf Nodes of Boosted Linear Trees

### ISSUE THREAD

**Issue by ToddMeng** [microsoft/LightGBM#6630]
## Summary
Enhancing the Flexibility of Linear Models in Leaf Nodes of Boosted Linear Trees

## Motivation

Linear trees represent a practical technique that not only enhances model performance and simplifies model structure but also improves model interpretability. When working with linear models, users often need to impose numerous custom constraints to enhance interpretability and incorporate additional prior knowledge. These constraints may include restricting all regression coefficients to be positive, defining the monotonicity of each variable, and limiting the linear regression to a subset of selected features.

## Description

As a regular user of this library, I am deeply grateful for the diligent efforts of all developers and maintainers, whose hard work has greatly facilitated our work.
Upon a thorough review of the documentation and the linear_tree_learner.cpp code (link: https://github.com/microsoft/LightGBM/blob/master/src/treelearner/linear_tree_learner.cpp), I have observed that, apart from the ridge regression parameters, the linear model component lacks support for other features, such as the aforementioned constraints on the signs of regression coefficients and the capability to include only a subset of features in the linear regression.

## References
It is proposed that the functionality extensions of linear models in sklearn could be referenced, or an interface could be provided to enable users to customize linear models, thereby enhancing the flexibility and practicality of linear tree models.

**Comment by jaguerrerod** [comment]
Related with this I think adding the option to include some predictors in all linear models, besides predictors used in splits to reach the leaf, is important.
I have datasets that have data of several segments of population and I'm not interested in include the variables define the segments in the model but include an adjustment in prediction using the segment flags in the linear model fitted on each leaf.
My leaves have more than 20K observations so including this adjustment by segment hasn't overfitting problem.
This option could be set by a parameter 'features_forced_to_leaf_linear_model' as an array of index of features or feature names.
I think this isn't complex to implement but I haven't the skills in C++ to do it.

**Comment by tblume1992** [comment]
Summary
When linear_tree=True, LightGBM fits a ridge regression at each leaf using
all numeric features. It would be valuable to let users specify which
features are eligible for the leaf linear model, while still allowing all
features to participate in splits.

Motivation
Time series: AR lag features may benefit from linear leaves
(extrapolation beyond training range) while
calendar, promo, etc. might not. Currently if we try to overcome extrapolation with LightGBM using linear_tree=True we overfit significantly.
Domain knowledge often identifies
which features have linear structure and which don't, currently it is all or nothing.
Description
Proposed API
Two potential options:

# Option A — extend the linear_tree parameter to accept a feature list/indices
dtrain = lgb.Dataset(
    X, label=y,
    params={"linear_tree": ["lag_1", "lag_7", "lag_14", "lag_28"]},
)

# Option B — separate parameter for selectivity (keeps linear_tree boolean)
dtrain = lgb.Dataset(
    X, label=y,
    params={
        "linear_tree": True,
        "linear_tree_features": ["lag_1", "lag_7", "lag_14", "lag_28"],
    },
)
LigthtGBM typically dominates demand forecasting competitions (and my personal and professional experience) such as VN1 and M5 but extrapolation is still a natural issue with trees and time series.

This issue was raised years ago so I am resurfacing it: https://github.com/lightgbm-org/LightGBM/issues/6630

Thanks so much!

Moved here
@jameslamb
